optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -305,6 +477,11 @@ def prepare_inputs_for_generation(
                 input_ids = input_ids[:, remove_prefix_length:]
             else:
                 input_ids = torch.index_select(input_ids, 1, token_idx - 1)
+        elif reuse_cache and token_idx is not None:
+            # With reuse_cache, KV cache is pre allocated hence for the 1st token we can
+            # slice the inputs till token idx for the fwd pass
+            input_ids = input_ids[:, :token_idx]
+            attention_mask = attention_mask[:, :token_idx]
```

Comment:
Hi @atakaha , I ran the following instructions with and without 'reuse_cache' enabled and got different outputs:
```shell
python run_generation.py --model_name_or_path mosaicml/mpt-7b --use_hpu_graphs --use_kv_cache --limit_hpu_graph --batch_size 8 --max_input_tokens 128 --max_new_tokens 128 --trim_logits --attn_softmax_bf16 --warmup 3 --n_iterations 3 --bf16
```
```shell
python run_generation.py --model_name_or_path mosaicml/mpt-7b --use_hpu_graphs --use_kv_cache --limit_hpu_graph --batch_size 8 --max_input_tokens 128 --max_new_tokens 128 --trim_logits --attn_softmax_bf16 --warmup 3 --n_iterations 3 --bf16 --reuse_cache
```
Then I found these lines of code will cause different behavior with the original code. Specifically, the shape of 'input_ids' will change from (batchsize, head_num, **max_input_tokens+max_new_tokens**, head_dim) to  (batchsize, head_num, **max_input_tokens**, head_dim). The 'input_tensor' will be padded at [here](https://github.com/atakaha/optimum-habana/blob/d7d53288b26c11430ac6fbbf0b5e94c2f08e33f2/optimum/habana/transformers/generation/utils.py#L883). 

When I comment out these lines, I get the same output as the version with 'reuse_cache' disabled.

I'm really confused because this implementation is very similar to line 479. Anyway, I will continue to understand the logic.

Reply:
Hi @Jing1Ling , Thank you for find this issue. and appreciate any further comments/inputs. I also revise this.

Reply:
I guess this is a precision issue. With 'reuse_cache' disabled, I tested different 'max_input_tokens', which will change the length of padding, and got inconsistent output.
I mentioned before that I got same output when I comment several lines of code. The logic of the commented codes is slicing the input when decode the first token, which also change the length of padding.
Although the padding will be masked and shouldn't affect the output, the actual implementation is to set a minimum float value for it. Then after softmax operation, the padded position will have a small probability close to zero, so the length of padding may affect the output.
And when I use fp32, all outputs under the following settings seem to be consistent:
	• disable reuse_cahce, different max_input_tokens
	• enable reuse_cache
	• enbale reuse_cache and bucket_internal, different bucket internal(also affect length of padding)
However, I didn't save the log yesterday and I can't book a Gaudi machine for confirmation at the moment. It may need a double-check.

Reply:
I'm observing key_states precision difference between with/without reuse_cache.